### ISSUE TITLE
S4487 FP with ??= operator.

### ISSUE THREAD

**Issue by markusschaber** [SonarSource/sonar-dotnet#6653]
### Description

I get an S4487 false positive with the ??= operator, the equivalent code using `if` does not trigger the rule.

### Repro steps
```
public class Class1
{
    /// <summary>
    /// Our delay to debounce when there are lots of rapid changes to the same file.
    /// </summary>
    private static readonly TimeSpan DebounceDelay = TimeSpan.FromSeconds(1);

    private readonly object _lock = new();

    // This line trigger the rule
    private Task? _checkerTask;

    private bool _dirty;

    public void TriggerBackgroundChecks()
    {
        lock (_lock)
        {
            _dirty = true;
            // start a new checker task when there's none running.
            _checkerTask ??= BackgroundCheckLoopAsync();
        }
    }

    /* This code with if does not trigger the rule. */
    // public void TriggerBackgroundChecksWithIf()
    // {
    //     lock (_lock)
    //     {
    //         _dirty = true;
    //         if (_checkerTask == null)
    //             _checkerTask = BackgroundCheckLoopAsync();
    //     }
    // }

    private async Task BackgroundCheckLoopAsync()
    {
        while (true)
        {
            lock (_lock)
            {
                if (!_dirty)
                {
                    // unregister the checker task and exit.
                    _checkerTask = null;
                    return;
                }

                _dirty = false;
            }

            // Perform some checks...
            await PerformExpensiveChecksAsync();

            // wait for the delay before we run the next check.
            await Task.Delay(DebounceDelay);
        }
    }

    private ValueTask PerformExpensiveChecksAsync() => ValueTask.CompletedTask;
}

```
### Expected behavior

As the variable is semantically important in this case (I only want a single Checker Task to run at the same time), removing the member as suggested by the rule makes no sense. So I expect the rule to not trigger in this example.

### Actual behavior

The rule triggers on the variable _checkerTask, and claims it's "unread", and I should remove it.

### Known workarounds

A workaround is to use the commented TriggerBackgroundChecksWithIf() method, which is semantically valid. But my IDE (JetBrains Rider) then suggests that I simplify the `if` statement to the equivalent `??=` assignment.

### Related information

* C#/VB.NET Plugins version: SonarLint für Rider 7.3.0.59206
* JetBrains Rider: 2022.3.1 Build #RD-223.8214.53
* MSBuild / dotnet version: .NET Sdk 7.0.100
* Operating System: Windows 10

**Comment by Corniel** [comment]
I can confirm this issue (with an even way smaller code sample)

``` C#
class NullCoalescingAssignment
{
    private object field;

    public void SetIfNull()
    {
        field ??= new object(); // Noncompliant FP
    }
}
```

**Comment by Corniel** [comment]
I would argue it is the other way around: In the first case, a field is set but never used. In the second one, the value only checked to be set, but never to be used. So it is a FN op the second piece of code.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @markusschaber,

Thank you for reporting this. I'll mark it as FP, as the field is indeed read and used to take a decision. Semantically, `??=` is a syntax sugar for this, and it should behave the same
```
_checkerTask = _checkerTask ?? Task.Run(() => BackgroundCheckLoopAsync());
```

**Comment by Corniel** [comment]
I would argue it is the other way around: In the first case, a field is set but never used. In the second one, the value only checked to be set, but never to be used. So it is a FN op the second piece of code.

**Comment by pavel-mikula-sonarsource** [comment]
I saw your comment @Corniel :) I consider the field to be used for a state check, therefore read.

**Comment by Corniel** [comment]
@pavel-mikula-sonarsource Sorry. My browser suggested I did not posted yet. ;)

I understand your reasoning. However, in this snippet, if only assign something not to be used again, you should normally not keep that reference alive. So you should refactor it, just using a bool to check if you triggered the func. But that is indeed more complex to keep track of.

**Comment by markusschaber** [comment]
Hmm, as long as the Task is not prematurely cleaned up by the GC, I could really use a bool member instead.

My first Idea was to use the Task.IsCompleted property, instead of setting the field back to null. Initializing it with Task.CompletedTask would even allow me to mark the member as non-nullable, but this introduces a small race condition between leaving the loop and the task actually being completed, as the Task internal state machine does not honor our locks.
